vscode-remote-ssh: Use $VSCODE_AGENT_FOLDER instead of ~/.vscode-server to unbreak insiders

sliedes opened the pull request:
This fixes remote ssh on vscode insiders where the path is
~/.vscode-server-insiders instead of ~/.vscode-server. (#371686)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [x] `nix-build -A nixosTests.vscode-remote-ssh`. This tests the functionality on vanilla (non-insiders) vscode.
  - [x] Manually verified that the change fixes the issue on vscode-insiders.
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sliedes committed:
vscode-remote-ssh: Use $VSCODE_AGENT_FOLDER instead of ~/.vscode-server

This fixes remote ssh on vscode insiders where the path is
~/.vscode-server-insiders instead of ~/.vscode-server. (#371686)

sliedes commented on pkgs/applications/editors/vscode/extensions/ms-vscode-remote.remote-ssh/default.nix:
```diff
@@ -67,12 +67,12 @@ let
     ${lib.optionalString useLocalExtensions ''
       # Use local extensions
       if [ -d $HOME/.vscode/extensions ]; then
```
Note that here we still refer to `$HOME/.vscode/extensions` instead of, conditionally, `$HOME/.vscode-insiders/extensions`. I think this is a bit tricky decision in some ways: non-insiders generally uses `~/.vscode` and insiders `~/.vscode-insiders`. However, the fact that a user locally uses insiders by no means seems to require that they also use insiders on the remote when working on it locally.

In any case, IIUC:

- This "use local extensions" is a nixpkgs enhancement not supported by upstream vscode?
- This PR still does not break any use case by not changing this, and does fix remote SSH on insiders. If `useLocalExtensions`, it would use the extensions from a local non-insiders version of vscode, which IMO may or may not be what the user wants (either option seems right in some cases and wrong in some), but probably should generally work.

So I just left this as is.

drupol reviewed (commented):
I have no clue if this is OK, I'll wait a bit for some feedback.

sliedes commented:
> I have no clue if this is OK, I'll wait a bit for some feedback.

No great hurry from my side; while it'd be nice to get it merged, I manage this also well with a local patch, and I'd also love to get my first nixpkgs PR right. :D

Would it help if I explained more about the context (which must be familiar to the maintainer of the extension)? Although all this is deduced from reading the source and trying out things.

---

# Situation prior to my patch

The remote SSH extension runs the "vscode server", on the remote host. That server is implemented in JavaScript and is run using node.

I believe Microsoft has had enough problems with random node.js versions, so the extension downloads a node binary from the net, obviously requiring patching.

Vscode normally downloads the node binary to `~/.vscode-server/bin/node`. The patch in the derivation applies patchelf to that binary.

# My change

The only wart with this is that, well, the remote SSH in this package never worked for me. After some debugging I realized it's because I use vscode-insiders, the preview version of vscode. In that case the correct directory is `~/.vscode-server-insiders` instead of `~/.vscode-server`. Yet prior to my PR the patch hardcodes the path.

The path is passed to the patched JavaScript file. Unfortunately it doesn't seem to be documented what variables are passed to that script, which is minified and apparently closed source. But the original patch had found a good place to patch, the variables have English names, and `$VSCODE_AGENT_FOLDER` seems to be the correct variable. I feel much more confident after reading the test case which actually tests that vscode (non-insiders) can use this extension.

drupol commented:
Thank you for the details.

I've got a question, how sure can you be that the variable `$VSCODE_AGENT_FOLDER` exist? What if it does not?

sliedes commented:
[+@Enzime as the original author of the Nix test for this. I really love the work you put in the test!]

> I've got a question, how sure can you be that the variable `$VSCODE_AGENT_FOLDER` exist? What if it does not?

That is a reasonable question. I first interpreted it as how do I know the variable is set, but also what I think you meant (if the directory exists) is a good question.

`extension.js` is originally a nice little file with 611 kB of minified Javascript in one line. What the patch (already before my modifications) patches is actually a string inside that JavaScript that contains an embedded shell script. In the original it has `\n` instead of newlines, but apparently JavaScript allows either, so the patch adds real newlines.

I formatted the Javascript to see what the part with the shell script ends up looking like; you can find that part below (and, of course, we cannot guarantee that this won't change in some version). Search for `// NOTE` to find places I marked as relevant.

## What would happen if the environment variable is unset

I don't know what the `getScriptStartMarker` function returns; looking at the code, I hope it does execute it the shell `-u` to error on unset variables. If not and `$VSCODE_AGENT_FOLDER` would not be set, I guess what would end up trying to copy stuff in `/bin`. Relevant snippet:

```sh
serverDir="$VSCODE_AGENT_FOLDER/bin/$COMMIT_ID"  # // NOTE this is the (only) line I changed
serverNode="$serverDir/node"
echo "VS Code Node: $serverNode"

# Check if Node included with VS Code Server runs
if ! nodeVersion=$($serverNode -v); then
echo "VS Code Node Version: $nodeVersion"

if ! nix-build "<nixpkgs>" -A patchelf --out-link "$serverDir/patchelf" || ! "$serverDir/patchelf/bin/patchelf" --version; then
  echo "Failed to get patchelf from nixpkgs"
fi

if [ -e $serverNode.orig ]; then
  cp $serverNode.orig $serverNode
else
  cp $serverNode $serverNode.orig
fi
```

## What would happen if the directory `$VSCODE_AGENT_FOLDER` does not exist

The original (unpatched) script downloads a `node` binary to `$VSCODE_AGENT_FOLDER/bin/$COMMIT_ID/node`. Reading the code, it looks like we could use `$VSCH_DIR/node` instead. Of course neither is robust to relevant upstream changes in the script logic.

If it doesn't exist, the script will try to `patchelf` it regardless (this is in fact what I saw happen without my patch); the errors from this will end up in a vscode debug log.

After that the extension will execute the `node` from wherever it downloaded it—what happened to me without my patch is that the script tried to `patchelf` `~/.vscode-server/$COMMIT/bin/node` and failed, and then the extension executed `~/.vscode-server-insiders/$COMMIT/bin/node`. This will fail because it hasn't been patched to run on NixOS, again showing as errors in the vscode debug log.

## What makes me 20x more confident

All this feels quite fragile, and that must have been what @Enzime thought when writing a test for it (big kudos!): https://github.com/NixOS/nixpkgs/blob/fb74a6f8dcae79aa451377667eca7868608b3049/nixos/tests/vscode-remote-ssh.nix

The test looks like it starts two VMs, runs vscode (non-insiders) and this extension and checks that one of them is able to connect to the other. Of course if we wanted that could be improved to test the same on vscode-insiders, but 1) I suspect this test is not the cheapest test; 2) I would consider regular vscode the 50x more important case;  and 3) I think this working on that version gives us good confidence anyway that it works.

## The part of the js with the script (after formatting)

```js
                      const w = u ? u.map((e) => `--install-extension ${e}`).join(" ") : "";
                      return `
echo "${a.getScriptStartMarker(
                          e
                      )}"

#
# Setup
#
VSC_TMP="\${XDG_RUNTIME_DIR:-\`echo /tmp\`}"
COMMIT_ID="${c}"
SERVER_APP_NAME="${h}"
EXTENSIONS="${w}"
TELEMETRY="--telemetry-level ${l}"
export VSCODE_AGENT_FOLDER=${  # // NOTE
                          d ?? "$HOME"
                      }/${p}
ALLOW_CLIENT_DOWNLOAD="${"off" !== y ? 1 : 0}"
FORCE_CLIENT_DOWNLOAD="${"always" === y ? 1 : 0}"
now=\`date +%s\`

PORT_FLAG=${
                          E ? "" : S ? `--port=${S.start}-${S.end}` : "--port=0"
                      }

SERVER_LISTEN_FLAG=${
                          E ? '--socket-path="$VSC_TMP/vscode-ssh-remote-server-sock-$now"' : ""
                      }
UNPACK_RESULT=
DID_LOCAL_DOWNLOAD=0
DOWNLOAD_TIME=
INSTALL_TIME=
EXT_INSTALL_TIME=
SERVER_START_TIME=
CONNECTION_TOKEN=

VSCH_DIR="$VSCODE_AGENT_FOLDER/bin/$COMMIT_ID"
VSCH_SERVER_SCRIPT="$VSCH_DIR/bin/$SERVER_APP_NAME"
VSCH_LOGFILE="$VSCODE_AGENT_FOLDER/.$COMMIT_ID.log"
VSCH_PIDFILE="$VSCODE_AGENT_FOLDER/.$COMMIT_ID.pid"
VSCH_TOKENFILE="$VSCODE_AGENT_FOLDER/.$COMMIT_ID.token"

fail_with_exitcode() {
    echo "${e}: start"
    echo "exitCode==$1=="
    echo_common_results
    echo "${e}: end"
    exit 0
}

echo_common_results() {
    echo "listeningOn==$LISTENING_ON=="
    echo "osReleaseId==$OSRELEASEID=="
    echo "arch==$ARCH=="
    echo "tmpDir==$VSC_TMP=="
    echo "platform==$PLATFORM=="
    echo "unpackResult==$UNPACK_RESULT=="
    echo "didLocalDownload==$DID_LOCAL_DOWNLOAD=="
    echo "downloadTime==$DOWNLOAD_TIME=="
    echo "installTime==$INSTALL_TIME=="
    echo "extInstallTime==$EXT_INSTALL_TIME=="
    echo "serverStartTime==$SERVER_START_TIME=="
}


millis() {
    if [[ $PLATFORM == linux ]]; then
        echo $(($(date +%s%N)/1000000))
    else
        # Can't get actual millis on mac
        echo $(($(date +%s)*1000))
    fi
}

elapsed() {
    echo $(($(millis) - $1))
}

if [[ ! -d $VSCH_DIR ]]; then
    mkdir -p $VSCH_DIR

    if (( $? > 0 ))
    then
        echo "Creating the server install dir failed..."
        fail_with_exitcode ${
                          o.InstallExitCode.CreateInstallDirFailed
                      }
    fi
fi

printenv_indent() {
    echo "printenv:"
    printenv | sed 's/^/    /'
}



#
# Get OS name
#
OSRELEASEID=\`cat /etc/os-release 2>/dev/null | grep -a -E '^ID=' | sed 's/^[Ii][Dd]=//g' | sed 's/"//g'\`
if [[ -z $OSRELEASEID ]]
then
    OSRELEASEID=\`cat /usr/lib/os-release 2>/dev/null | grep -a -E '^ID=' | sed 's/^[Ii][Dd]=//g' | sed 's/"//g'\`
    if [[ -z $OSRELEASEID ]]
    then
        OSRELEASEID=\`uname -s\`
    fi
fi


#
# Get host platform/architecture
#
UNAME=$(uname -s)
case $UNAME in
    Linux) PLATFORM=linux;;
    Darwin) PLATFORM=macOS;;
    *)
        echo "Unsupported platform: $UNAME"
        fail_with_exitcode ${
                          o.InstallExitCode.UnsupportedPlatform
                      }
        ;;
esac

ARCH=$(uname -m)
case $ARCH in
    x86_64) VSCODE_ARCH="x64";;
    armv7l | armv8l)
        VSCODE_ARCH="armhf"
        ;;
    arm64 | aarch64)
        VSCODE_ARCH="arm64"
        ;;
    *)
        OSRELEASE=$(uname -r)
        case $OSRELEASE in
            *x86_64*) VSCODE_ARCH="x64";;
            *)
                echo "Unsupported architecture: $ARCH"
                fail_with_exitcode ${
                          o.InstallExitCode.UnsupportedArch
                      }
            ;;
        esac
        ;;
esac


#
# Verify linux prereqs
#
if [[ $PLATFORM == linux ]]; then
    if [[ -f /usr/lib64/libstdc++.so.6 ]]; then
        # Typical path
        libstdc_path='/usr/lib64/libstdc++.so.6'
    elif [[ -f /sbin/ldconfig ]]; then
        # Look up path
        libstdc_path=\`/sbin/ldconfig -p | grep 'libstdc++' | sed 's/^.* => '//i\`
    else
        echo "Warning: Can't find libstdc++.so or ldconfig, can't verify libstdc++ version"
    fi

    if [[ ! -z $libstdc_path ]]; then
        # can't use "strings", not included by default on some distros
        found_glibcxx=0
        reqd_glibcxx=304018 # 3.4.18
        glibcxx_versions=\`cat $libstdc_path | grep -a -o -E 'GLIBCXX_[[:digit:]]+\\.[[:digit:]]+\\.([[:digit:]]+)?' | sed 's/GLIBCXX_'//gi\`
        while read -r version; do
            if [[ $version =~ ([[:digit:]]+)\\.([[:digit:]]+)(\\.([[:digit:]]+))? ]]; then
                major="\${BASH_REMATCH[1]}"
                minor="\${BASH_REMATCH[2]}"
                patch="\${BASH_REMATCH[4]}"

                total=$(( $major*100000 + $minor*1000 ))
                if [[ ! -z $patch ]]; then
                    total=$(( $total + $patch ))
                fi

                if (( $total >= $reqd_glibcxx )); then
                    found_glibcxx=1
                    break
                fi
            fi
        done <<< "$glibcxx_versions"

        if [[ $found_glibcxx == 0 ]]; then
            echo "Missing GLIBCXX >= 3.4.18!"
            echo "Found versions $glibcxx_versions"
        fi
    fi
    reqd_glibc=217 # 2.17
    found_glibc=0
    ldd_version=\`ldd --version | head -n1\`
    if [[ $ldd_version =~ ^ldd.*([[:digit:]]+)\\.([[:digit:]]+) ]]; then
        major="\${BASH_REMATCH[1]}"
        minor="\${BASH_REMATCH[2]}"

        total=$(( $major*100 + $minor ))
        if (( $total >= $reqd_glibc )); then
            found_glibc=1
        fi
    fi

    if [[ $found_glibc == 0 ]]; then
        echo "Missing GLIBC >= 2.17!"
        echo "Found version $ldd_version"
    fi

    if [[ $found_glibc == 0 || $found_glibcxx == 0 ]]; then
        echo '${e}$$1$$'
    fi
fi

#
# Acquire install lock
#
MAIN_LOCKFD=99
CLEANUP_LOCKFD=98
${(function (
                          e,
                          t
                      ) {
                          return `

    use_flock=${
                              t ? "1" : "0"
                          }
    if [[ $PLATFORM == "macOS" ]]; then
        use_flock=0
    fi

    ## Copyright (C) 2009 Przemyslaw Pawelczyk <przemoc@gmail.com>
    ## This script is licensed under the terms of the MIT license.
    ## https://opensource.org/licenses/MIT

    get_lockfile() {
        lockfiles_in_tmp=${
                              e ? "1" : "0"
                          }
        lockfile_name=vscode-remote-lock.$USER.$1
        if [[ $lockfiles_in_tmp == "1" ]]; then
            echo "$VSC_TMP/$lockfile_name"
        else
            echo "$VSCODE_AGENT_FOLDER/bin/$1/$lockfile_name"  # // NOTE
        fi
    }

    # PRIVATE
    _lock() { flock -$1 $2; }

    _no_more_locking() {
        if [[ $use_flock == "1" ]]; then
            _lock u $2
            _lock xn $2 && rm -f $(get_lockfile $1)
        else
            lockfile_name=$(get_lockfile $1)
            rm -f $lockfile_name
            rm -f $lockfile_name.target
        fi
    }

    # PUBLIC - all take lock FD

    setup_locking() {
        if [[ $use_flock == "1" ]]; then
            eval "exec $2>\\"$(get_lockfile $1)\\"";
            trap "_no_more_locking $1 $2" EXIT
        fi
    }

    setup_lock_cleanup()  {
        if [[ $use_flock == "0" ]]; then
            trap "_no_more_locking $1" EXIT
        fi
    }

    # obtain an exclusive lock immediately or fail
    exlock_now() {
        if [[ $use_flock == "1" ]]; then
            _lock xn $1
        else
            lockfile_name=$(get_lockfile $2)
            touch $lockfile_name.target
            echo \\\\ln $lockfile_name.target $lockfile_name
            \\ln $lockfile_name.target $lockfile_name
        fi
    }

    # drop a lock
    unlock() {
        if [[ $use_flock == "1" ]]; then
            _lock u $1
        else
            _no_more_locking $2
        fi
    }
`;
                      })(
                          g,
                          v
                      )}
setup_locking $COMMIT_ID $MAIN_LOCKFD

echo "Acquiring lock on \`get_lockfile $COMMIT_ID\`"
exlock_now $MAIN_LOCKFD $COMMIT_ID
if (( $? > 0 ))
then
    echo "Installation already in progress..."
    echo "If you continue to see this message, you can try toggling the remote.SSH.useFlock setting"
    fail_with_exitcode ${
                          o.InstallExitCode.AlreadyInProgress
                      }
fi

setup_lock_cleanup $COMMIT_ID $MAIN_LOCKFD

#
# Delete old servers if needed
#
TO_DELETE=$(ls -1 -t $VSCODE_AGENT_FOLDER/bin | tail -n +6)
for COMMIT_TO_DELETE in $TO_DELETE; do
    echo "Found old VS Code install $COMMIT_TO_DELETE, attempting to clean up"

    setup_locking $COMMIT_TO_DELETE $CLEANUP_LOCKFD
    exlock_now $CLEANUP_LOCKFD $COMMIT_TO_DELETE
    if (( $? == 0 )); then
        setup_lock_cleanup $COMMIT_TO_DELETE $CLEANUP_LOCKFD
        RUNNING="\`ps ax | grep $COMMIT_TO_DELETE | grep -v grep | wc -l | tr -d '[:space:]'\`"
        if [[ $RUNNING == 0 ]]; then
            echo "Deleting old install from $VSCODE_AGENT_FOLDER/bin/$COMMIT_TO_DELETE"
            rm -rf $VSCODE_AGENT_FOLDER/bin/$COMMIT_TO_DELETE

            echo "Deleting old log file from $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.log"
            rm $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.log
            if [[ -f $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.pid ]]; then
                rm $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.pid
            fi
            if [[ -f $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.token ]]; then
                rm $VSCODE_AGENT_FOLDER/.$COMMIT_TO_DELETE.token
            fi
        else
            echo "Install still has running processes, not deleting: $COMMIT_TO_DELETE"
        fi
    else
        echo "Failed to acquire lock for install, not deleting: $COMMIT_TO_DELETE"
    fi
done

do_client_download() {
    DID_LOCAL_DOWNLOAD=1
    echo "Trigger local server download"
    echo ${n.getDownloadServerStartTrigger(
                          e
                      )}
    echo platform==$PLATFORM==
    echo vscodeArch==$VSCODE_ARCH==
    echo destFolder==$VSCODE_AGENT_FOLDER/bin/==
    echo destFolder2==$COMMIT_ID==
    echo ${n.getDownloadServerEndTrigger(
                          e
                      )}
    echo "Waiting for client to transfer server archive..."
    echo "Waiting for $VSCH_DIR/vscode-scp-done.flag and vscode-server.tar.gz to exist"

    while true; do
        if [[ -f $VSCH_DIR/vscode-scp-done.flag ]]; then
            if [[ ! -f $VSCH_DIR/vscode-server.tar.gz ]]; then
                echo "Found flag but not server tar - server transfer failed"
                fail_with_exitcode ${
                          o.InstallExitCode.ServerTransferFailed
                      }
            fi

            echo "Found flag and server on host"
            rm $VSCH_DIR/vscode-scp-done.flag
            break
        else
            echo -n ' '
            sleep 3
        fi
    done
}

do_client_download_or_fail() {
    if [[ $DID_LOCAL_DOWNLOAD == "1" ]]; then
        echo "Already attempted local download, failing"
        fail_with_exitcode $1
    elif [[ $ALLOW_CLIENT_DOWNLOAD == "1" ]]; then
        do_client_download
        LOCAL_DOWNLOAD_RESULT=success
    else
        fail_with_exitcode $1
    fi
}

do_host_download() {
    start=$(millis)
    echo '${i.getProgressOutputString(
                          e,
                          1
                      )}'

    if [[ $PLATFORM == linux ]]; then
        PLATFORM_DOWNLOAD_PATH=server-linux-$VSCODE_ARCH
    elif [[ $VSCODE_ARCH == "arm64" ]]; then
        PLATFORM_DOWNLOAD_PATH=server-darwin-arm64
    else
        PLATFORM_DOWNLOAD_PATH=server-darwin
    fi

    DOWNLOAD_URL=https://update.code.visualstudio.com/commit:$COMMIT_ID/$PLATFORM_DOWNLOAD_PATH${
                          m ? "-web" : ""
                      }/${r}

    wget --version &> /dev/null
    if (( $? == 0 ))
    then
        echo "Downloading with wget"
        WGET_ERRORS=$(2>&1 wget ${
                          ((b = y), "off" !== b ? "--tries=1" : "")
                      } --connect-timeout=7 --dns-timeout=7 -nv -O vscode-server.tar.gz $DOWNLOAD_URL)
        if (( $? != 0 ))
        then
            echo "wget download failed"
            echo $WGET_ERRORS
            printenv_indent

            do_client_download_or_fail ${
                          o.InstallExitCode.ServerDownloadFailed
                      }
        else
            echo "Download complete"
            DOWNLOAD_TIME=$(elapsed $start)
        fi
    else
        curl --version &> /dev/null
        if (( $? == 0 ))
        then
            echo "Downloading with curl"
            CURL_OUTPUT=$(curl --connect-timeout 7 -L $DOWNLOAD_URL --output vscode-server.tar.gz -w "%{http_code}")
            if [[ ($? != 0) || ($CURL_OUTPUT != 2??) ]]
            then
                echo "curl download failed"
                echo $CURL_OUTPUT
                printenv_indent

                do_client_download_or_fail ${
                          o.InstallExitCode.ServerDownloadFailed
                      }
            else
                echo "Download complete"
                DOWNLOAD_TIME=$(elapsed $start)
            fi
        else
            printenv_indent
            echo "Neither wget nor curl is installed"

            do_client_download_or_fail ${
                          o.InstallExitCode.NoDownloaderAvailable
                      }
        fi
    fi
}

do_install() {
    start=$(millis)
    echo '${i.getProgressOutputString(
                          e,
                          2
                      )}'
    UNPACK_RESULT=success
    echo 'tar --version:'
    tar --version
    tar -xf vscode-server.tar.gz
    TAR_EXIT=$?
    INSTALL_TIME=$(elapsed $start)

    if (( $TAR_EXIT > 0 ))
    then
        echo "ERROR: tar exited with non-0 exit code: $?"
        UNPACK_RESULT=error
        do_client_download_or_fail ${
                          o.InstallExitCode.UnpackFailed
                      }
        do_install
        return
    fi

    mv vscode-server*/* .

    # cheap sanity check
    if [[ ! -f $VSCH_DIR/node || ! -f $VSCH_SERVER_SCRIPT ]]  # // NOTE node existence check
    then
        UNPACK_RESULT=missingFiles
        echo "ERROR: $VSCH_DIR/node or $VSCH_SERVER_SCRIPT don't exist"
        do_client_download_or_fail ${
                          o.InstallExitCode.UnpackFailed
                      }
        do_install
        return
    fi

    # Delete the leftover folder which might have -web prefix, and the .tar.gz
    rm -rf vscode-server*
}

do_extension_install() {
    # install extensions
    if [[ ! -z $EXTENSIONS ]]
    then
        echo "Installing extensions..."
        start=$(millis)
        $VSCH_SERVER_SCRIPT $TELEMETRY $EXTENSIONS
        EXT_INSTALL_TIME=$(elapsed $start)
    fi
}

#
# Install if needed
#
if [[ ! -f $VSCH_SERVER_SCRIPT ]]
then
    echo "Installing to $VSCH_DIR..."
    STASHED_WORKING_DIR="\`pwd\`"
    cd $VSCH_DIR

    if [[ $FORCE_CLIENT_DOWNLOAD == "1" ]]; then
        do_client_download
    else
        do_host_download
    fi

    do_install
    cd $STASHED_WORKING_DIR
else
    echo "Found existing installation at $VSCH_DIR..."
fi


#
# Generate a new connection token
#
write_connection_token() {
    rm $VSCH_TOKENFILE
    CONNECTION_TOKEN="${s.generateUuid()}"
    touch $VSCH_TOKENFILE
    chmod 0700 $VSCH_TOKENFILE
    echo $CONNECTION_TOKEN > $VSCH_TOKENFILE
}

#
# Use Node from nixpkgs for NixOS hosts
#

serverDir="$VSCODE_AGENT_FOLDER/bin/$COMMIT_ID"  # // NOTE this is the (only) line I changed
serverNode="$serverDir/node"
echo "VS Code Node: $serverNode"

# Check if Node included with VS Code Server runs
if ! nodeVersion=$($serverNode -v); then
echo "VS Code Node Version: $nodeVersion"

if ! nix-build "<nixpkgs>" -A patchelf --out-link "$serverDir/patchelf" || ! "$serverDir/patchelf/bin/patchelf" --version; then
  echo "Failed to get patchelf from nixpkgs"
fi

if [ -e $serverNode.orig ]; then
  cp $serverNode.orig $serverNode
else
  cp $serverNode $serverNode.orig
fi

if ! nix-build "<nixpkgs>" -A bintools --out-link $serverDir/bintools; then
  echo "Failed to build bintools from nixpkgs"
fi

INTERPRETER=$(cat $serverDir/bintools/nix-support/dynamic-linker)

echo "Interpreter from bintools: $INTERPRETER"

if ! nix-build "<nixpkgs>" -A stdenv.cc.cc.lib --out-link $serverDir/cc; then
  echo "Failed to build stdenv.cc.cc.lib from nixpkgs"
fi

if ! $serverDir/patchelf/bin/patchelf --set-interpreter $INTERPRETER --set-rpath $serverDir/cc-lib/lib $serverNode; then
  echo "Failed to patch Node binary"
fi

rm "$serverDir/patchelf"
fi

nodeVersion=$($serverNode -v)
echo "VS Code Node Version: $nodeVersion"

if ! nodeVersion=$($serverNode -v); then
echo "Unable to fix Node binary, quitting"
fail_with_exitcode ${o.InstallExitCode.ServerTransferFailed}
fi



#
# Start the server
#
start_server() {
    echo "Starting server with command... $VSCH_SERVER_SCRIPT --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown $PORT_FLAG $TELEMETRY $SERVER_LISTEN_FLAG $EXTENSIONS &> \\"$VSCH_LOGFILE\\" < /dev/null"
    printenv_indent
    start=$(millis)
    write_connection_token
    if [[ -f $VSCH_LOGFILE ]]; then
        echo "Removing old logfile at $VSCH_LOGFILE"
        rm $VSCH_LOGFILE # See #6265
    fi
    $VSCH_SERVER_SCRIPT --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown $PORT_FLAG $TELEMETRY $SERVER_LISTEN_FLAG $EXTENSIONS --connection-token-file $VSCH_TOKENFILE &> "$VSCH_LOGFILE" < /dev/null &
    echo $! > "$VSCH_PIDFILE"
    echo "Spawned remote server: $!"

    stopTime=$((SECONDS+8))

    while (($SECONDS < $stopTime))
    do
        LISTENING_ON=\`cat "$VSCH_LOGFILE" | grep -a -E 'Extension host agent listening on .+' | grep -v grep | sed 's/Extension host agent listening on //'\`
        if [[ $LISTENING_ON != '' ]]
        then
            break
        fi

        echo "Waiting for server log..."
        sleep .03
    done

    SERVER_START_TIME=$(elapsed $start)
}

echo "Checking $VSCH_LOGFILE and $VSCH_PIDFILE for a running server"
if [[ -f $VSCH_PIDFILE ]]; then
    SERVER_PID=\`cat $VSCH_PIDFILE\`
    echo "Looking for server with pid: $SERVER_PID"
    ALL_PROCS=\`ps ax | grep "^\\s*$SERVER_PID\\s"\`
else
    ALL_PROCS=\`ps ax\`
fi

RUNNING="\`echo $ALL_PROCS | grep $VSCH_SERVER_SCRIPT | grep -v grep | wc -l | tr -d '[:space:]'\`"
if [[ $RUNNING == 0 ]]; then
    start_server
elif [[ ! -f $VSCH_LOGFILE ]]; then
    echo "WARNING: VS Code Server is running but its logfile is missing. Don't delete the VS Code Server manually, run the command 'Uninstall VS Code Server'."
    start_server
else
    echo "Found running server..."
    do_extension_install
    if [[ -f $VSCH_TOKENFILE ]]; then
        CONNECTION_TOKEN=\`cat $VSCH_TOKENFILE\`
    fi
fi

echo " "
echo "*"
echo "* Reminder: You may only use this software with Visual Studio family products,"
echo "* as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)"
echo "*"
echo " "

LISTENING_ON=\`cat "$VSCH_LOGFILE" | grep -a -E 'Extension host agent listening on .+' | grep -v grep | sed 's/Extension host agent listening on //'\`
WEBUITOKEN=\`cat "$VSCH_LOGFILE" | grep -a -E 'Web UI available at' | grep -o -E '${
                          t.webUITokenParameter
                      }=[0-9a-f-]+' | grep -o -E '[0-9a-f-]+'\`
if [[ -z $LISTENING_ON ]]
then
    echo "Server did not start successfully. Full server log at $VSCH_LOGFILE >>>"
    cat $VSCH_LOGFILE
    echo "<<< End of server log"
    fail_with_exitcode ${
                          o.InstallExitCode.ServerFailedToStart
                      }
fi

#
# If the server was already running, ensure that it won't shut down in the near future
#
if [[ $RUNNING != 0 ]]; then
    # In case the server is close to timing out and shutting down, ask it to delay shutdown

    wget --version &> /dev/null
    if [[ $? == 0 && $LISTENING_ON =~ (^[[:digit:]]+$) ]]
    then
        echo "Checking server status on port $LISTENING_ON with wget"
        WGET_OUT=$(wget --no-proxy -nv -O - http://127.0.0.1:$LISTENING_ON/delay-shutdown 2>&1)
        if [[ $? != 0 ]]; then
            echo $WGET_OUT
            fail_with_exitcode ${
                          o.InstallExitCode.StatusCheckFailed
                      }
        fi
    else
        curl --version &> /dev/null
        if (( $? == 0 ))
        then
            if [[ $LISTENING_ON =~ (^[[:digit:]]+$) ]]
            then
                echo "Checking server status on port $LISTENING_ON with curl"
                CURL_OUTPUT=$(curl --noproxy 127.0.0.1 -s http://127.0.0.1:$LISTENING_ON/delay-shutdown -w " %{http_code}")
            else
                echo "Checking server status on socket $LISTENING_ON with curl"
                CURL_OUTPUT=$(curl --noproxy 127.0.0.1 -s --unix-socket $LISTENING_ON http://127.0.0.1/delay-shutdown -w " %{http_code}")
            fi

            if [[ ($? != 0) || ($CURL_OUTPUT != "OK 200") ]]; then
                echo $CURL_OUTPUT
                fail_with_exitcode ${
                          o.InstallExitCode.StatusCheckFailed
                      }
            fi
        else
            echo "Neither wget nor curl is installed"
            fail_with_exitcode ${
                          o.InstallExitCode.NoDownloaderAvailable
                      }
        fi
    fi
fi


# What we echo below cannot be wider than 80 characters
echo "${e}: start"
${_.map((e) => `echo ${e}==$${e}==`).join(
                          "
"
                      )}
echo "webUiAccessToken==$WEBUITOKEN=="
echo_common_results
echo "connectionToken==$CONNECTION_TOKEN=="
echo "${e}: end"

unlock $MAIN_LOCKFD $COMMIT_ID

${
                          f ? "while true; do sleep 180; echo -n ' '; done" : ""
                      }
`;
                      var b;
                  });
          },
          6523: (e, t, r) => {
              "use strict";
```

drupol commented on pkgs/applications/editors/vscode/extensions/ms-vscode-remote.remote-ssh/default.nix:
```diff
@@ -67,12 +67,12 @@ let
-        if [ -e $HOME/.vscode-server/extensions ]; then
-          mv $HOME/.vscode-server/extensions $HOME/.vscode-server/extensions.bak
+        if [ -e $VSCODE_AGENT_FOLDER/extensions ]; then
+          mv $VSCODE_AGENT_FOLDER/extensions $VSCODE_AGENT_FOLDER/extensions.bak
```
Why do we need to keep `extensions.bak` ?

sliedes replied:
I honestly have no idea; I tried to keep the change minimal, only changing `$HOME/.vscode-server` to `$VSCODE_AGENT_FOLDER`. This "Use local extensions feature" isn't something I use myself at all, so I'm wary of making more sweeping changes to it.
